hivelink-core/src/main/java/org/apache/iceberg/hivelink/core/schema/MergeHiveSchemaWithAvro.java

```diff
@@ -143,10 +144,14 @@ public Schema union(UnionTypeInfo union, Schema partner, List<Schema> results) {
   @Override
   public Schema primitive(PrimitiveTypeInfo primitive, Schema partner) {
     boolean shouldResultBeOptional = partner == null || AvroSchemaUtil.isOptionSchema(partner);
+    int nullIndex = 0;
+    if (partner != null && AvroSchemaUtil.isOptionSchema(partner)) {
+      nullIndex = AvroSchemaUtil.getNullIndex(partner);
+    }
     Schema hivePrimitive = hivePrimitiveToAvro(primitive);
     // if there was no matching Avro primitive, use the Hive primitive
     Schema result = partner == null ? hivePrimitive : checkCompatibilityAndPromote(hivePrimitive, partner);
-    return shouldResultBeOptional ? AvroSchemaUtil.toOption(result) : result;
+    return shouldResultBeOptional ? AvroSchemaUtil.toOption(result, nullIndex == 1) : result;
```

Comment:
This handling is incomplete, it only handles cases for primitives inside containers (e.g. list), but does not handle similar cases for nested lists, maps, structs, etc. E.g. this case will fail
```
  @Test
  public void shouldNotReorderListElementType1() {
    String hive = "struct<fa:array<array<int>>>";
    Schema avro =
            SchemaBuilder.record("r1")
                    .fields()
                    .name("fa")
                    .type()
                    .nullable()
                    .array()
                    .items()
                    .nullable()
                    .array()
                    .items()
                    .nullable()
                    .intType()
                    .arrayDefault(Collections.singletonList(Arrays.asList(1, 2, 3)))
                    .endRecord();

    assertSchema(avro, merge(hive, avro));
  }
  ```